libvast/src/system/source_command.cpp

```diff
@@ -41,13 +68,29 @@ int reader_command_base::run_impl(caf::actor_system& sys,
   // Helper for blocking actor communication.
   scoped_actor self{sys};
   // Spawn the source.
-  auto src_opt = make_source(self, options, begin, end);
+  auto src_opt = make_source(self, options);
   if (!src_opt) {
-    std::cerr << "unable to spawn source: " << sys.render(src_opt.error())
-              << std::endl;
+    VAST_ERROR(self, "was not able to spawn a source",
+               sys.render(src_opt.error()));
     return EXIT_FAILURE;
   }
   auto src = std::move(*src_opt);
+  // Supply an alternate schema, if requested.
+  if (auto sf = caf::get_if<std::string>(&options, "schema")) {
+    auto schema = load_schema_file(*sf);
+    if (schema)
+      self->send(src, put_atom::value, std::move(*schema));
+    else
+      VAST_ERROR(self, "had a schema error:", sys.render(schema.error()));
+  }
+  // Attempt to parse the remainder as an expression.
+  if (begin != end) {
+    auto expr = parse_expression(begin, end);
+    if (expr)
+      self->send(src, std::move(*expr));
+    else
+      VAST_ERROR(self, sys.render(expr.error()));
+  }
```

Comment:
I would say that both an invalid schema or expression should result in a "hard" error. That is, the function should return an error and not just display one. POLS again: the user expects that everything went well, but the server side (which might be not on their radar) only records the error in the log file.

Reply:
Good point, I second that. 👍

Reply:
Well said.